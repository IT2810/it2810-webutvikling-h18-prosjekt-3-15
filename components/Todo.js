import React, { Component } from 'react';
import {
    View,
    Text,
    StyleSheet,
    TextInput,
    ScrollView,
    Button,
    TouchableOpacity, 
    AsyncStorage
} from 'react-native';
import Task from './Task.js';
import CountdownComponent from "./CountdownComponent.js";

export default class Todo extends Component {
    constructor(props) {
        super(props);
        this.state = {
            taskDict: {},
            keyCount: 0,
            taskText: '',
        };
        this.deleteTask = this.deleteTask.bind(this);

    }

    async getTaskDict(){
        try{
            let value = await AsyncStorage.getItem('taskdict');
            if(value !== null){
                console.log('value : ' + value);
                this.setState(JSON.parse(value))
            }

        } catch (error){
            console.log("Error retrieving data " + error);
        }
    }

    async saveTaskDict(value){
        try{
            await AsyncStorage.setItem('taskdict', JSON.stringify(this.state));
            console.log("save value : " + JSON.stringify(value));
        }catch(error){
            console.log("Error saving data " + error);
        }
    }

    componentDidMount(){
        this.getTaskDict()
    }

    render() {
        let taskDict = this.state.taskDict;
        const nav = this.props.navigation;
        return (
            <View>
                <View style={styles.footer}>
                    <TextInput
                        style={styles.textInput}
                        placeholder='> Type your task here and press the +'
                        onChangeText={(taskText)=> this.setState({taskText})}
                        value={this.state.taskText}
                        placeholderTextColor='white'
                        underlineColorAndroid='transparent'>
                    </TextInput>
                </View>
                <Text style={styles.backgroundInput}>
                </Text>
                <TouchableOpacity onPress={ this.addTask.bind(this) } style={styles.addButton}>
                    <Text style={styles.addButtonText}>+</Text>
                </TouchableOpacity>
                <View style={styles.scrollViewContainer}>
                    <ScrollView>
                        {(Object.keys(taskDict)).map((key) => {
                            return <Task key={key} taskKey={key} task={taskDict[key]["taskText"]} date={taskDict[key]["date"]} navigation={nav}
                                         checked={taskDict[key]['checked']}
                                         deleteMethod = {() => this.deleteTask(key)}
                                         callback={() => this.checkedBox(key)}/>} )
                        }
                    </ScrollView>
                </View>
            </View>
        );
    }

    addTask(){
        if(this.state.taskText){
            let keyInt = this.state.keyCount +1;
            var d = new Date();
            let newDict = this.state.taskDict;
            newDict[keyInt] = {
                'taskText': this.state.taskText,
                    'date': d.getFullYear() + "/" + (d.getMonth()+1) + "/"+ d.getDate(),
                'checked': false
            };
            this.setState({
                ...this.state,
                keyCount: keyInt,
                taskDict: newDict,
                taskText:''
            });
            this.saveTaskDict(newDict);
        }
    }

    deleteTask(key){
        let newDict = this.state.taskDict;
        delete newDict[key];
        this.setState({
            ...this.state,
            taskDict: newDict});
    }

    checkedBox(key){
        let taskDictA = this.state.taskDict;
        taskDictA[key]['checked'] = ! taskDictA[key]['checked'];
        this.setState({
            ...this.state,
            taskDict: taskDictA
        })
    }

}
const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    scrollViewContainer:{
        height: 550,
        marginTop:70,
    },
    footer: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        zIndex: 10
    },
    textInput: {
        alignSelf: 'stretch',
        color: '#fff',
        padding: 20,
        marginRight: 70,
        backgroundColor: '#252525',
    },
    addButton: {
        position: 'absolute',
        zIndex: 11,
        right: 5,
        top: 3,
        backgroundColor: '#2fc47c',
        width: 60,
        height: 60,
        borderRadius: 35,
        alignItems: 'center',
        justifyContent: 'center',
        elevation: 8
    },
    backgroundInput: {
        position: 'absolute',
        zIndex: 10,
        right: 0,
        top: 0,
        backgroundColor: '#252525',
        width: 80,
        height: 68,
        alignItems: 'center',
        justifyContent: 'center',
        elevation: 8
    },
    addButtonText: {
        color: '#fff',
        fontSize: 24
    }
});
