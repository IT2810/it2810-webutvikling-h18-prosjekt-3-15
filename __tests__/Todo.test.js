import React from 'react';
import Todo from '../components/Todo.js';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
    const tree = renderer
    // Quick workaround for navigation. Without this it has trouble finding the state and therefore its params and so on
        .create(<Todo navigation={{'state':{'params':{'url': '/1/test/'}}}}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
it('check start state', ()=>{
    const tree = renderer.create(<Todo navigation={{'state':{'params':{'url': '/1/test/'}}}}/>).getInstance();
    let x = tree.state;
    console.log(x);
    expect(Object.is({taskDict: {}, keyCount: 0, taskText: '', urlKey: '/1/test/'}, x))
})
