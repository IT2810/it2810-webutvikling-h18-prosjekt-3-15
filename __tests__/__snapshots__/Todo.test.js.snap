// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View>
    <View
      style={
        Object {
          "marginTop": 7,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderRadius": 5,
                    "justifyContent": "center",
                    "marginHorizontal": 2,
                  },
                  Object {
                    "backgroundColor": "#4286f4",
                  },
                  Object {
                    "height": 78,
                    "width": 69,
                  },
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontVariant": Array [
                        "tabular-nums",
                      ],
                      "fontWeight": "bold",
                    },
                    Object {
                      "fontSize": 30,
                    },
                    Object {
                      "color": "#000",
                    },
                  ]
                }
              >
                00
              </Text>
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "white",
                  "marginVertical": 2,
                },
                Object {
                  "fontSize": 16.666666666666668,
                },
                Object {
                  "color": "#000",
                },
              ]
            }
          >
            Minutes
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "borderRadius": 5,
                    "justifyContent": "center",
                    "marginHorizontal": 2,
                  },
                  Object {
                    "backgroundColor": "#4286f4",
                  },
                  Object {
                    "height": 78,
                    "width": 69,
                  },
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "white",
                      "fontVariant": Array [
                        "tabular-nums",
                      ],
                      "fontWeight": "bold",
                    },
                    Object {
                      "fontSize": 30,
                    },
                    Object {
                      "color": "#000",
                    },
                  ]
                }
              >
                05
              </Text>
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "white",
                  "marginVertical": 2,
                },
                Object {
                  "fontSize": 16.666666666666668,
                },
                Object {
                  "color": "#000",
                },
              ]
            }
          >
            Seconds
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 15,
            "marginRight": 15,
          },
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          pointerEvents="box-only"
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#9E9E9E",
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 19,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              Object {
                "padding": 12,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {},
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Array [
                  Object {
                    "color": "white",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontSize": 17.5,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            Play/Pause
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 162,
        "zIndex": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="> Type your task here and press the +"
      placeholderTextColor="white"
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#252525",
          "color": "#fff",
          "marginRight": 70,
          "padding": 20,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#252525",
        "elevation": 8,
        "height": 68,
        "justifyContent": "center",
        "position": "absolute",
        "right": 0,
        "top": 162,
        "width": 80,
        "zIndex": 10,
      }
    }
  />
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#2fc47c",
        "borderRadius": 35,
        "elevation": 8,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 5,
        "top": 165,
        "width": 60,
        "zIndex": 11,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#fff",
          "fontSize": 24,
        }
      }
    >
      +
    </Text>
  </View>
  <View
    style={
      Object {
        "height": 400,
        "marginTop": 70,
      }
    }
  />
</View>
`;
